#!/usr/bin/python3
# -*- coding : utf-8 -*-
##########################################################
#
#       Filename:   fungi_from_gbk_get_cds_V1.0.py
#         Author:   yujie
#    Description:   fungi_from_gbk_get_cds_V1.0.py
#        Version:   1.0
#           Time:   2023/02/01 14:02:07
#  Last Modified:   2023/02/01 14:02:07
#        Contact:   hi@arcsona.cn
#        License:   GNU General Public License v3.0
#
##########################################################
from Bio import SeqIO
from Bio.Seq import Seq
# from icecream import ic
import argparse
# import linecache
import os
# import pretty_errors
import re
import sys
import time
# import copy
parser = argparse.ArgumentParser(
    add_help=False, usage='\n\
python3 cp_from_gbk_get_cds_V3.0.py -i [gbk dir] -o [out dir]\n\
Each species generates three files of cds,trna and complete sequence.')
optional = parser.add_argument_group('optional')
required = parser.add_argument_group('required')
required.add_argument(
    '-i', '--input',  help='gbk dir path', type=str)
required.add_argument(
    '-o', '--output',  help='output dir path', type=str)
optional.add_argument('-c', '--check', help='create genes via gap,not running by default,input "-c" when using',
                      action='store_true', required=False)
optional.add_argument('-d', '--duplicates', help='remove duplicates,default true',
                      action='store_false', required=False)
optional.add_argument('-ignore', '--ignore', help='ignore prompt,default true',
                      action='store_false', required=False)
optional.add_argument('-info', '--info', help='show update log and exit',
                      action='store_true', required=False)
optional.add_argument('-h', '--help', action='help',
                      help='show this help message and exit')
args = parser.parse_args()

if args.info:
    print('\næ›´æ–°æ—¥å¿—:')
    print('\t20220906 ä¿®æ”¹åŸºå› ç»„çš„ç±»å‹åˆ¤æ–­,è‹¥ä¸ç¬¦åˆåˆ™é€€å‡º')
    print('\t20221012 ä¿®æ”¹åŸºå› åå­—æ˜ å°„å‡½æ•°,ä½¿å…¶èƒ½å¤Ÿè¯†åˆ«CO1æ ·å¼')
    print('\t20221111 å¢åŠ gbké‡ŒåŸºå› åæ˜¯å…¨ç§°(å¦‚ATPase)æ—¶çš„å¤„ç†')
    print('\t20221217 feat: âœ¨ å¯¹gbkæ–‡ä»¶è¿›è¡Œå»é‡')
    print('\t20221219 ğŸfix(get_gene_note): ä¿®æ”¹æ²¡æœ‰geneæ ‡ç­¾çš„cdsçš„é»˜è®¤ID')
    print('\t20221220 âœ¨feat(main): å†™å…¥å»é‡åçš„ç™»å½•å·')
    print('\n')
    sys.exit(0)

# ############################################################################################


def format_fasta(note, seq, num):
    format_seq = ""
    for index, char in enumerate(seq):
        format_seq += char
        # if (index + 1) % num == 0:#å¯ä»¥ç”¨æ¥æ¢è¡Œ
        #format_seq += "\n"
    return note.strip() + "\n" + format_seq + "\n"


def ir(s):  # åå‘äº’è¡¥
    re = s[::-1]  # å­—ç¬¦ä¸²åå‘
    c = ""
    for i in re:
        if i == 'A':
            c = c + 'T'
        elif i == 'G':
            c = c + 'C'
        elif i == 'T':
            c = c + 'A'
        elif i == 'C':
            c = c + 'G'
    return c
# ###########################################################################################################


def get_complete_note(seq_record):  # è·å–æ•´ä¸ªå®Œæ•´åŸºå› ç»„ID
    try:
        seq_id = ''
        seq_record.description = seq_record.description.replace('/', '_')
        # 20220819 NC_044756.1.gbk voucher Liu HM/CP02 chloroplast  ç¬¬ä¸€æ¬¡å¤„ç†,å»æ‰ç‰¹æ®Šç¬¦å·
        # if seq_record.description.find('chloroplast'):#æœ‰bug,ç”¨stræ ¼å¼åŒ–åå°±æ²¡é—®é¢˜äº†
        # 20220627 if str(seq_record.description).find('chloroplast') -1ä¹Ÿæˆç«‹,åˆ¤æ–­æ—¶ä¸€å®šè¦ä»¥False Trueä¸ºå‡†
        # or seq_record.description.split(',')[-2].split()[-1] == 'chloroplast' or seq_record.description.split(',')[-2].split()[-1] == 'plastid':

        if str(seq_record.description).find('chloroplast') > 0 \
            or str(seq_record.description).find('plastid') > 0 \
                or seq_record.description.split(',')[-2].split()[-1] == 'chloroplast' \
                or seq_record.description.split(',')[-2].split()[-1] == 'plastid':  # gbk å±äº å¶ç»¿ä½“
            if str(seq_record.description).find('chloroplast') > 0:  # 20220819 æ ¹æ®ä¸åŒå…³é”®è¯åˆ†å‰²
                seq_id = seq_record.description.split(
                    'chloroplast')[0].replace(' ', '_').rstrip('_')
            elif str(seq_record.description).find('plastid') > 0:
                seq_id = seq_record.description.split(
                    'plastid')[0].replace(' ', '_').rstrip('_')
            name = seq_record.name
            if seq_id == name:
                seq_id = seq_id
            elif seq_id != name:
                seq_id = seq_id+'_'+name
            complete_note = ">" + seq_id + "\n"  # chloroplast--å¶ç»¿ä½“

        elif str(seq_record.description).find('mitochondrion') > 0 \
                or str(seq_record.description).find('mitochondrial') > 0:
            # 20220811 NC_031548.gbk æè¿°éƒ¨åˆ† DEFINITION  Oxynoemacheilus angorae mitochondrial DNA, complete genome.
            seq_id = seq_record.description.split(
                'mitochondrion')[0].replace(' ', '_').rstrip('_')  # å½¢å¦‚ UNVERIFIED:_Rumina_decollata
            if seq_id.startswith('UNVERIFIED:_'):  # ç¬¬äºŒæ¬¡å¤„ç†,å»æ‰ UNVERIFIED:_
                # seq_id = seq_id.lstrip('UNVERIFIED:_')  # æœ‰bug ???????
                seq_id = seq_id.split(':')[1].strip('_')
            # Cerion_watlingense_voucher_USNM:1514170_MN904501 ç¬¬ä¸‰æ¬¡å¤„ç†,å»æ‰å†’å·åå†…å®¹
            if len(seq_id.split(':')) > 1:
                seq_id = seq_id.split(':')[0]
            name = seq_record.name  # è¦ä¹ˆæ˜¯ç™»å½•å·  è¦ä¹ˆæ˜¯æ ·æœ¬
            if seq_id == name:
                seq_id = seq_id
            elif seq_id != name:
                seq_id = seq_id+'_'+name
            complete_note = ">" + seq_id + "\n"  # mitochondrion--çº¿ç²’ä½“

        else:
            print(seq_record.description)
            print('Genome Type WARNING! {}!\n{}'.format(
                seq_record.description.split(', ')[-2].split()[-1],
                'Program will automatically process!'))
            complete_note = ">" + (seq_record.description.split('mitochondrion')
                                   [0]).replace(' ', '_').rstrip('_') + "\n"

    except:  # å¦‚æœé‡åˆ°ä»»ä½•å‡ºé”™
        print('try/except')
        complete_note = ''
        #gbk_type = input('genome type(1:chloroplast;2:mitochondrion): ')
        gbk_type = 2
        if gbk_type == 1:
            seq_id = seq_record.description.split(
                'chloroplast')[0].replace(' ', '_').rstrip('_')  # ç‰©ç§æˆ–æ ·å“å
            if seq_id.startswith('UNVERIFIED:_'):  # å»æ‰ UNVERIFIED:_
                seq_id = seq_id.lstrip('UNVERIFIED:_')
            # å»æ‰Cerion_watlingense_voucher_USNM:1514170_MN904501 ä¸­ å†’å·åçš„å†…å®¹
            if len(seq_id.split(':')) > 1:
                seq_id = seq_id.split(':')[0]
            name = seq_record.name  # è¦ä¹ˆæ˜¯ç™»å½•å·  è¦ä¹ˆæ˜¯æ ·æœ¬
            if seq_id == name:
                seq_id = seq_id
            elif seq_id != name:
                seq_id = seq_id+'_'+name
            complete_note = ">" + seq_id + "\n"
        elif gbk_type == 2:
            print(seq_record.description)
            seq_id = input('seq_id: ')
            complete_note = ">" + seq_id + "\n"

    return complete_note, seq_id

# ###################################################################################################################
# ä»…åœ¨ get_gene_note(ele, complete_seq, seq_id, tmp_gene_name)ä¸­ä½¿ç”¨ï¼Œ
# ä»¥åº”å¯¹"gbkä¸­cdsæ²¡æœ‰/geneæ ‡ç­¾ï¼Œä½†æ˜¯æœ‰/productæ ‡ç­¾"çš„æƒ…å†µ


def gene_name_standardization_1(gene_name):  # çº¿ç²’ä½“ æ ¼å¼åŒ–åŸºå› åå­—,å¯é‡å¤ä½¿ç”¨
    all_gene_dict = {'ATP synthase F0 subunit 6': 'ATP6', 'ATP synthase F0 subunit 8': 'ATP8', 'cytochrome b': 'CYTB',
                     'cytochrome c oxidase subunit I': 'COX1', 'cytochrome c oxidase subunit II': 'COX2',
                     'cytochrome c oxidase subunit III': 'COX3', 'NADH dehydrogenase subunit 1': 'ND1',
                     'NADH dehydrogenase subunit 2': 'ND2', 'NADH dehydrogenase subunit 3': 'ND3',
                     'NADH dehydrogenase subunit 4': 'ND4', 'NADH dehydrogenase subunit 4L': 'ND4L',
                     'NADH dehydrogenase subunit 5': 'ND5', 'NADH dehydrogenase subunit 6': 'ND6'}
    if gene_name in all_gene_dict.keys():
        gene_name = all_gene_dict[gene_name]
    return gene_name


def gene_name_standardization_2(gene_name):  # æ ¼å¼åŒ–åŸºå› åå­—,å¯é‡å¤ä½¿ç”¨
    # åˆå§‹åŒ–
    all_gene_list_upper = ['ATP6', 'ATP8', 'CYTB', 'COX1', 'COX2',
                           'COX3', 'ND1', 'ND2', 'ND3', 'ND4', 'ND4L', 'ND5', 'ND6']
    all_gene_list_upper2 = ['ATP6', 'ATP8', 'CYTB', 'CO1', 'CO2',
                            'CO3', 'ND1', 'ND2', 'ND3', 'ND4', 'ND4L', 'ND5', 'ND6']
    all_gene_list_lower = ['atp6', 'atp8', 'cob', 'cox1', 'cox2',
                           'cox3', 'nad1', 'nad2', 'nad3', 'nad4', 'nad4l', 'nad5', 'nad6']
    name_flag = 0
    # åŸºå› åå­—å‰å¤„ç†1 # 20220825 gbkæ–‡ä»¶é‡Œ åŸºå› åæ˜¯ç½—é©¬æ•°å­—ï¼Œå› æ­¤éœ€è¦å…ˆå¤„ç†
    gene_name = gene_name.replace('III', '3').replace(
        'II', '2').replace('I', '1')
    # åŸºå› åå­—å‰å¤„ç†2 # 20221111 gbkæ–‡ä»¶é‡Œ åŸºå› åæ˜¯å…¨ç§° å¦‚'ATPase '
    gene_name = gene_name.replace('ATPase ', 'ATP').replace('Cyt ', 'CYT')

    # ------------------åå­—æ˜ å°„
    # åå­—æ ·å¼å¤§å†™çš„æƒ…å†µ
    if gene_name.upper() in all_gene_list_upper:
        gene_name = gene_name.upper()
    else:
        i = 0
        while i < 13:
            # 20221012  MN053900.1.gbk æœ‰ä¸€ä¸ªCO1
            if all_gene_list_upper2[i] == gene_name.upper():
                gene_name = all_gene_list_upper[i]
                print(gene_name)
                break
            else:
                i += 1
        if i >= 13:
            #print(gene_name, 'Try further changes...')
            # åå­—æ ·å¼å°å†™çš„æƒ…å†µ
            i = 0
            while i < 13:
                # 20220624  JN619347.gbk æœ‰ä¸€ä¸ªnad4L å¤§å°å†™æ··åˆ å°±ç¦»è°±
                if all_gene_list_lower[i] == gene_name.lower():
                    gene_name = all_gene_list_upper[i]
                    break
                else:
                    i += 1
            if i >= 13:
                if args.ignore == False:
                    print('all_three_methods_failed!')
                    print(gene_name, 'WARNING!Please check!')
                name_flag = 1
    return gene_name, name_flag

# ###################################################################################################################


def merge_sequence(ele, complete_seq):  # åˆå¹¶è·å–åˆ°çš„åºåˆ—,ç”¨äºå‡½æ•°(è·å–cds)æ¥è°ƒç”¨
    cds_seq = ""
    tmp_list = []  # ä½ç½®åˆ—è¡¨
    for ele1 in ele.location.parts:
        if ele1.strand == (-1):
            # print('minus')
            tmp_list.append(re.findall(
                r'\d+', str(ele1.end))[0])  # å®é™…èµ·ç‚¹,ä»endä¸­å–ä¸ç”¨+1
            tmp_list.append(str(int(re.findall(
                r'\d+', str(ele1.start))[0])+1))  # å®é™…ç»ˆç‚¹,ä»startå–+1
            cds_seq += ir(complete_seq[ele1.start:ele1.end])
        elif ele1.strand == (1):
            # print('plus')
            tmp_list.append(str(int(re.findall(
                r'\d+', str(ele1.start))[0])+1))  # å®é™…èµ·ç‚¹,è¦+1
            tmp_list.append(re.findall(
                r'\d+', str(ele1.end))[0])  # å®é™…ç»ˆç‚¹,ä¸ç”¨+1
            # åˆ‡ç‰‡æ²¡é—®é¢˜,ç´¢å¼•ä»startåˆ°end-1,ä¹Ÿå°±æ˜¯å¯¹åº”start+1åˆ°endçš„åºåˆ—
            cds_seq += complete_seq[ele1.start:ele1.end]
    return tmp_list, cds_seq


# ###################################################################################################################
def get_gene_note(file_no, file_name, ele, complete_seq, seq_id, tmp_gene_name):  # è·å–geneçš„idåŠåºåˆ—
    '''
    ä¼ å…¥ä¸Šä¸€ä¸ªåŸºå› ä¿¡æ¯tmp_gene_name,è¿”å›å½“å‰åŸºå› ä¿¡æ¯tmp_gene_name
    OrderedDict(
        [
        ('codon_start', ['1']),
        ('gene', ['rps19']),
        ('product', ['ribosomal protein S19']),
        ('protein_id', ['UKP82335.1']),
        ('transl_table', ['11']),
        ('translation', [
         'MTRSLKKNPFVANHLLRKINKLNTKAEKEIIITWSRASTIIPTMIGHTIAIHNGKEHLPIYITDRMVGHKLGEFSPTLNFRGHAKNDNRSRR'])
        ]
        )
    '''
    # 20220825 NC_034226.gbk cdsæ²¡æœ‰/geneæ ‡ç­¾ï¼Œä½†æ˜¯æœ‰/productæ ‡ç­¾
    if 'gene' not in ele.qualifiers.keys():  # è¿”å›ä¸Šä¸€ä¸ªåŸºå› ,å¥½ä»å…¶ä»–å‚è€ƒæ‰¾è¿™ä¸ªæ²¡åå­—çš„
        if 'product' in ele.qualifiers.keys():
            tmp_gene_name = ele.qualifiers['product'][0]
            tmp_gene_name = gene_name_standardization_1(tmp_gene_name)
        else:
            try:
                tmp_gene_name = tmp_gene_name+'_next'
            except:
                tmp_gene_name = input(
                    "\n{0}: {1} Previous: {2}. Current: {3}.\nPlease input current gene name:".
                    format(file_no, file_name, tmp_gene_name, ele.location.parts))

    elif 'gene' in ele.qualifiers.keys():
        tmp_gene_name = ele.qualifiers['gene'][0]

    tmp_gene_name, name_flag = gene_name_standardization_2(
        tmp_gene_name)  # 20221012 ä¹Ÿéœ€è¦ç¬¬äºŒç§æ ‡å‡†åŒ–

    tmp_list, gene_seq = merge_sequence(ele, complete_seq)
    pos_info_in_gene_note = " ["
    for i in range(2*len(ele.location.parts)):
        if i % 2 == 0:
            pos_info_in_gene_note += tmp_list[i]+'..'
        elif i % 2 == 1:
            pos_info_in_gene_note += tmp_list[i]+';'
    pos_info_in_gene_note = pos_info_in_gene_note.rstrip(';')+"]"
    gene_note = ">" + seq_id + pos_info_in_gene_note + " [gene=" + \
        tmp_gene_name + "]" + "\n"  # '>'åçš„æ ¼å¼å’Œå·²æœ‰è„šæœ¬å…¼å®¹
    return gene_note, gene_seq, tmp_gene_name

# ##############################################################################################


def get_gene(gbk_file_path, flag, dict_gene_len, file_no):  # è§£ægbkæ–‡ä»¶è·å–cds
    """å®Œæ•´åŸºå› ç»„"""
    file_name = os.path.basename(gbk_file_path)
    seq_record = SeqIO.read(gbk_file_path, "genbank")
    complete_seq = str(seq_record.seq)
    complete_note, seq_id = get_complete_note(seq_record)
    complete_fasta = format_fasta(complete_note, complete_seq, 70)  # 70æ¢è¡Œæœ¬ä¾‹ä¸é‡‡ç”¨
    """geneåºåˆ—"""
    cds_count = 0  # å¯¹cdsæ•°é‡è®¡æ•°
    trna_count = 0
    cds_fasta = ""
    trna_fasta = ""
    list_cds_name = []  # ç»Ÿè®¡cdsç§ç±»ï¼Œè®°å½•é‡å¤
    list_cds_name_len = []  # ç»Ÿè®¡cdsç§ç±»ï¼Œè®°å½•é‡å¤,è®°å½•é•¿åº¦
    list_trna_name = []  # ç»Ÿè®¡trnaç§ç±»,è®°å½•é‡å¤
    tmp_gene_name = ''  # ä¸Šä¸€ä¸ªåŸºå› åå­—,ä¸ºå­å‡½æ•°get_gene_note()å‡†å¤‡çš„
    for ele in seq_record.features:
        '''
        CDS
        tRNA
        rRNA
        repeat_region
        '''
        if ele.type == "CDS":
            cds_count += 1
            cds_note, cds_seq, tmp_gene_name = get_gene_note(
                file_no, file_name, ele, complete_seq, seq_id, tmp_gene_name)
            # list_cds_name.append(tmp_gene_name)  # æœ¬æ¬¡çš„åŸºå› åå­— å¤ç”¨,çº¿ç²’ä½“çš„è¯,åœ¨ä¸‹ä¸€éƒ¨åˆ†å­˜å…¥åˆ—è¡¨
            cds_fasta += format_fasta(cds_note, cds_seq, 70)
            gene_name = tmp_gene_name
            gene_name, name_flag = gene_name_standardization_2(gene_name)
            list_cds_name.append(gene_name)  # å­˜å…¥åˆ—è¡¨
            list_cds_name_len.append(
                gene_name+':'+str(len(cds_seq)))  # å­˜å…¥åˆ—è¡¨
            if name_flag == 1:
                print(gbk_file_path, 'WARNING!Please check!')
            dict_gene_len[gene_name].append(
                3*(len(ele.qualifiers['translation'][0])+1))  # cdsåºåˆ—é•¿åº¦
            if (flag):  # eleæœ‰å¯èƒ½æ˜¯trna,è¦ç¡®ä¿å…ˆæ‰¾åˆ°ä¸€ä¸ªcdsåæ‰èƒ½é€€å‡º,æ‰€ä»¥æ”¾ä¸Šé¢ifçš„ä¸‹ä¸€çº§
                break
        elif ele.type == 'tRNA':
            trna_count += 1
            trna_note, trna_seq, tmp_gene_name = get_gene_note(
                file_no, file_name, ele, complete_seq, seq_id, tmp_gene_name)
            trna_fasta += format_fasta(trna_note, trna_seq, 70)
            gene_name = tmp_gene_name
            gene_name = gene_name  # gene_name_standardization(gene_name)
            list_trna_name.append(gene_name)  # å­˜å…¥åˆ—è¡¨

    s = '{2}: {0} has {1} CDS'.format(file_name, cds_count, file_no)
    if cds_count == 0:
        # --------There may be no comments--------'.format(
        s = '{2}: {0} has {1} CDS'.format(file_name, cds_count, file_no)
        print(s.ljust(50), '----------There may be no comments----------')
    elif cds_count != 0:
        # 20220811 è¾“å‡ºå·¦å¯¹é½  str.ljust(50)  è¾¾åˆ°50ä¸ªå­—ç¬¦å®½åº¦
        if args.ignore == False:
            print(s.ljust(50), '+', list_cds_name_len)
        else:
            print(s.ljust(50), '+', list_cds_name_len)
    return file_name, seq_id, complete_fasta, cds_fasta, cds_count, list_cds_name,  trna_fasta, trna_count, list_trna_name, dict_gene_len, s, list_cds_name_len

# ###############################################################################æ„é€ å­å‡½æ•°


def create_gene_by_gap(dict_missing_gene, dict_gene_len, out_cds_file_list):  # ç”¨gapæ„é€ æ²¡æœ‰çš„åŸºå› 
    # dict_missing_gene   '>Achatina_fulica_NC_024601': []
    # out_cds_file_list   'cds_NC_036381.fasta'
    out_cds_file_list_1 = []
    for i in dict_missing_gene.keys():
        cds_fasta = ''
        for j in dict_missing_gene[i]:
            ave = round(sum(dict_gene_len[j]) /
                        len(dict_gene_len[j]))  # è¯¥åŸºå› å¹³å‡é•¿åº¦
            cds_note = (i+' [0..0]'+' [gene={}]').format(j)
            cds_seq = ave*'-'
            cds_fasta += format_fasta(cds_note, cds_seq, 70)
        for m in out_cds_file_list:
            m = m.lstrip('cds_').rstrip('.fasta')
            if i.find(m) >= 0 or i.find(m.rstrip('.1')) >= 0:  # ié‡Œå¤´ä¹Ÿå¯èƒ½ä¸å¸¦ç‰ˆæœ¬å·,å»æ‰mçš„å†æŸ¥æ‰¾
                cds_file_path = 'cds_'+m+'.fasta'
                out_cds_file_list_1.append(cds_file_path)

        with open(args.output+os.sep+cds_file_path, 'ab+') as f_cds:  # è¯»å†™æ‰“å¼€ä¸€ä¸ªäºŒè¿›åˆ¶æ–‡ä»¶ï¼Œå…è®¸è¯»ï¼Œæˆ–åœ¨æ–‡ä»¶æœ«è¿½åŠ æ•°æ®
            f_cds.write(cds_fasta.encode())


if __name__ == '__main__':
    #################################################################
    # æ ¼å¼åŒ–æˆ2016-03-20 11: 45: 39å½¢å¼
    begin_time = time.time()
    start_time = time.strftime('%Y-%m-%d %H:%M:%S', time.localtime())
    print('Start Time : {}'.format(start_time))
    #################################################################
    print('\n')
    if not os.path.exists(args.output):
        os.makedirs(args.output)
    if (not os.path.exists(os.path.join(args.output, 'trna'))):
        os.makedirs(os.path.join(args.output, 'trna'))
    if (not os.path.exists(os.path.join(args.output, 'cds'))):
        os.makedirs(os.path.join(args.output, 'cds'))
    if (not os.path.exists(os.path.join(args.output, 'complete'))):
        os.makedirs(os.path.join(args.output, 'complete'))

    """å†™å…¥ç»Ÿè®¡æ–‡ä»¶"""
    with open((args.output+os.sep+'log'), 'w') as f_log:
        f_log.write(
            'gene{0}atp6{0}atp8{0}cob{0}cox1{0}cox2{0}cox3{0}nad1{0}nad2{0}nad3{0}nad4{0}nad4L{0}nad5{0}nad6\n'.format('\t'))  # å°å†™
        f_log.write(
            'gene{0}ATP6{0}ATP8{0}CYTB{0}COX1{0}COX2{0}COX3{0}ND1{0}ND2{0}ND3{0}ND4{0}ND4L{0}ND5{0}ND6\n'.format('\t'))  # å¤§å†™
    all_gene_list_upper = ['ATP6', 'ATP8', 'CYTB', 'COX1', 'COX2',
                           'COX3', 'ND1', 'ND2', 'ND3', 'ND4', 'ND4L', 'ND5', 'ND6']
    all_gene_list_lower = ['atp6', 'atp8', 'cob', 'cox1', 'cox2',
                           'cox3', 'nad1', 'nad2', 'nad3', 'nad4', 'nad4l', 'nad5', 'nad6']

    """ç»Ÿè®¡åˆå§‹åŒ–"""
    dict_missing_gene = {}  # æ¯ä¸ªæ–‡ä»¶ä¸­ç¼ºå¤±çš„åŸºå› ç»Ÿè®¡,æ€» å­—å…¸
    dict_gene_len = {}  # ç»Ÿè®¡æ¯ä¸ªåŸºå› åœ¨ä¸åŒç‰©ç§ä¸­çš„é•¿åº¦,å–å¹³å‡
    list_seq_id_de_duplication = []  # å»é‡çš„ç‰©ç§
    list_unique_accession = []  # å»é‡çš„ç™»å½•å·
    for i in all_gene_list_upper:
        dict_gene_len[i] = []

    """åˆå§‹åŒ–"""
    dict_file_cds_count = {}  # æ¯ä¸ªæ–‡ä»¶ä¸­cdsè®¡æ•°
    file_list = [x for x in os.listdir(
        args.input) if (os.path.isfile(os.path.join(args.input, x)) and x.endswith('gbk'))]
    file_list.sort()  # key=lambda x: int(x.split('.')[0])) #æ ¹æ®æ–‡ä»¶åä¸­çš„æ•°å­—

    """ä¸»ç¨‹åº"""
    file_no = 0  # ä¸ºäº†å±å¹•è¾“å‡ºæ›´ç›´è§‚,åœ¨æ–‡ä»¶å‰é¢çš„ç¼–å·
    out_cds_file_list = []  # æ‰€æœ‰è¾“å‡ºçš„cdså‰ç¼€æ–‡ä»¶è·¯å¾„,ç”¨äºç”¨GAPæ„é€ åºåˆ—
    for file in file_list:
        file_no += 1
        gbk_file_path = os.path.join(args.input, file)
        file_name, seq_id, complete_fasta, cds_fasta, cds_count, list_cds_name,  \
            trna_fasta, trna_count, list_trna_name, dict_gene_len, s, list_cds_name_len = get_gene(
                gbk_file_path, False, dict_gene_len, file_no)
        dict_file_cds_count[file_name] = cds_count  # æ¯ä¸ªæ–‡ä»¶ä¸­cdsè®¡æ•°
        '''
        20221217å¯¹gbkæ–‡ä»¶å»é‡
        '''
        seq_id_content = seq_id.split('_')
        if seq_id.find('NC_') > 0:
            species = '_'.join(seq_id_content[:-2])
            accession = '_'.join(seq_id_content[-2:])
        else:
            species = '_'.join(seq_id_content[:-1])
            accession = '_'.join(seq_id_content[-1:])
        if args.duplicates == True:  # æ‰§è¡Œå»é‡
            if species not in list_seq_id_de_duplication:
                list_seq_id_de_duplication.append(species)
                list_unique_accession.append(accession)

                """å†™å…¥æ–‡ä»¶"""
                out_cds_file = ('cds_' + file_name.rstrip('.gbk')+'.fasta')
                out_cds_file_list.append(out_cds_file)  # gapæ„é€ åŸºå› 
                out_complete_file = (seq_id+'.fasta')
                out_trna_file = ('trna_'+file_name.rstrip('.gbk')+'.fasta')
                out_log_file = 'log'
                with open((os.path.join(args.output, 'complete', out_complete_file)), 'wb') as f_complete, \
                    open((os.path.join(args.output, 'cds', out_cds_file)), 'wb') as f_cds,\
                        open(os.path.join(args.output, 'trna', out_trna_file), 'wb') as f_trna,\
                        open(os.path.join(args.output,  out_log_file), 'a+') as f_log:
                    f_complete.write(complete_fasta.encode())
                    f_cds.write(cds_fasta.encode())
                    f_trna.write(trna_fasta.encode())

                    """ä»¥ä¸‹ä¸ºç»Ÿè®¡éƒ¨åˆ†"""
                    f_log.write(s+'\n')
                    f_log.write('>'+file.rstrip('.gbk')+'\t')
                    list_missing_gene = []  # æ¯ä¸ªæ–‡ä»¶ä¸­ç¼ºå¤±çš„åŸºå› ç»Ÿè®¡,å•åˆ—è¡¨
                    for i in range(len(all_gene_list_upper)):
                        if all_gene_list_upper[i] in list_cds_name \
                                or all_gene_list_upper[i].lower() in list_cds_name \
                                or all_gene_list_lower[i] in list_cds_name \
                                or all_gene_list_lower[i].upper() in list_cds_name:
                            f_log.write(all_gene_list_upper[i]+'\t')
                        else:
                            f_log.write('NULL'+'\t')
                            list_missing_gene.append(
                                all_gene_list_upper[i])  # ç¼ºå¤±çš„åŸºå› 
                    if 1 <= len(list_missing_gene) < 13:
                        print(''.ljust(55), '-', list_missing_gene)

                    f_log.write('\n')
                    [f_log.write(tmp+'\t') for tmp in list_missing_gene]
                    f_log.write('\n')
                dict_missing_gene['>'+seq_id] = list_missing_gene  # æ”¾å…¥å­—å…¸å½“ä¸­

        else:

            """å†™å…¥æ–‡ä»¶"""
            out_cds_file = ('cds_' + file_name.rstrip('.gbk')+'.fasta')
            out_cds_file_list.append(out_cds_file)  # gapæ„é€ åŸºå› 
            out_complete_file = (seq_id+'.fasta')
            out_trna_file = ('trna_'+file_name.rstrip('.gbk')+'.fasta')
            out_log_file = 'log'
            with open((os.path.join(args.output, 'complete', out_complete_file)), 'wb') as f_complete, \
                open((os.path.join(args.output, 'cds', out_cds_file)), 'wb') as f_cds,\
                    open(os.path.join(args.output, 'trna', out_trna_file), 'wb') as f_trna,\
                    open(os.path.join(args.output,  out_log_file), 'a+') as f_log:
                f_complete.write(complete_fasta.encode())
                f_cds.write(cds_fasta.encode())
                f_trna.write(trna_fasta.encode())

                """ä»¥ä¸‹ä¸ºç»Ÿè®¡éƒ¨åˆ†"""
                f_log.write(s+'\n')
                f_log.write('>'+file.rstrip('.gbk')+'\t')
                list_missing_gene = []  # æ¯ä¸ªæ–‡ä»¶ä¸­ç¼ºå¤±çš„åŸºå› ç»Ÿè®¡,å•åˆ—è¡¨
                for i in range(len(all_gene_list_upper)):
                    if all_gene_list_upper[i] in list_cds_name \
                            or all_gene_list_upper[i].lower() in list_cds_name \
                            or all_gene_list_lower[i] in list_cds_name \
                            or all_gene_list_lower[i].upper() in list_cds_name:
                        f_log.write(all_gene_list_upper[i]+'\t')
                    else:
                        f_log.write('NULL'+'\t')
                        list_missing_gene.append(
                            all_gene_list_upper[i])  # ç¼ºå¤±çš„åŸºå› 
                if 1 <= len(list_missing_gene) < 13:
                    print(''.ljust(55), '-', list_missing_gene)

                f_log.write('\n')
                [f_log.write(tmp+'\t') for tmp in list_missing_gene]
                f_log.write('\n')
            dict_missing_gene['>'+seq_id] = list_missing_gene  # æ”¾å…¥å­—å…¸å½“ä¸­

    # åé¢è€ƒè™‘åœ¨è¿™ç»™å­—å…¸æ’ä¸ªåº
    with open((args.output+os.sep+'log'), 'a+') as f_log:
        f_log.write(str(dict_missing_gene))
    """é™¤äº†ç‰©ç§1å¤–,å…¶ä»–ç‰©ç§æ‰€æœ‰cdsæ€»æ•°"""
    total_ref_gene = 0
    for i in dict_file_cds_count.keys():  # é”®ä¸ºseq_id,å€¼ä¸ºä¸ªæ•°
        total_ref_gene += dict_file_cds_count[i]
    print('\n')
    # print(dict_gene_len)  # é”®ä¸ºæ¯ä¸ªåŸºå› ,å€¼ä¸ºåˆ—è¡¨,åˆ—è¡¨ä¸ºæ¯ä¸ªåŸºå› åœ¨ä¸åŒç‰©ç§ä¸­çš„é•¿åº¦
    # print(dict_missing_gene)  # é”®ä¸º>seq_id,å€¼ä¸ºåˆ—è¡¨,åˆ—è¡¨ä¸ºæ¯ä¸ªç‰©ç§ç¡®å®çš„åŸºå› 
    # print(2*(total_ref_gene-13))
    # ic(len(dict_missing_gene))
    # ic(dict_missing_gene)
    # ic(len(out_cds_file_list))
    """gapæ„é€ åŸºå› """
    if args.check:
        create_gene_by_gap(dict_missing_gene, dict_gene_len, out_cds_file_list)

    if len(list_unique_accession) != 0:
        print('\n{} left after removing duplicates\n'.format(
            len(list_unique_accession)))
    else:
        print('\nall species are reserved\n')

    with open((os.path.join(args.output, 'list_tre')), 'w') as list_handle:
        for i in list_unique_accession:
            list_handle.write(i+'\n')
    ###############################################################
    end_time = time.strftime('%Y-%m-%d %H:%M:%S', time.localtime())
    print('End Time : {}'.format(end_time))
    print('Already Run {}s'.format(time.time()-begin_time))
    print('Done')
    ###############################################################
